Fatras/include/ActsFatras/Kernel/InteractionList.hpp

```diff
@@ -0,0 +1,344 @@
+  /// @param[in]     processIndex is the index of the process to be executed
+  /// @param[in,out] particle     is the particle being updated
+  /// @param[out]    generated    is the container of generated particles
+  /// @return Break condition, i.e. whether a process killed the particle
+  ///
+  /// The process index is expected to originate from a previous
+  /// `armPointLike(...)` call, but this is not enforced. How to select the
+  /// correct process requires more information that is not available here.
+  template <typename generator_t>
+  bool runPointLike(generator_t& rng, size_t processIndex, Particle& particle,
+                    std::vector<Particle>& generated) const {
+    return runPointLikeImpl(rng, processIndex, particle, generated,
+                            PointLikeIndices());
+  }
+
+ private:
+  // allow processes to be masked. defaults to zeros -> no masked processes
+  Mask m_mask;
+  Processes m_processes;
+
+  // for the `runContinuous` call, we need to iterate over all available
+  // processes and apply the ones that implement the continuous process
+  // interface. this is done using an index-based compile-time recursive call.
+  template <typename generator_t, std::size_t kI0, std::size_t... kIs>
+  bool runContinuousImpl(generator_t& rng, const Acts::MaterialSlab& slab,
+                         Particle& particle, std::vector<Particle>& generated,
+                         std::index_sequence<kI0, kIs...>) const {
+    const auto& process = std::get<kI0>(m_processes);
+    // only call process if it is not masked
+    if (not m_mask[kI0] and process(rng, slab, particle, generated)) {
+      // exit early in case the process signals an abort
+      return true;
+    }
+    return runContinuousImpl(rng, slab, particle, generated,
+                             std::index_sequence<kIs...>());
+  }
+  template <typename generator_t>
+  bool runContinuousImpl(generator_t&, const Acts::MaterialSlab&, Particle&,
+                         std::vector<Particle>&, std::index_sequence<>) const {
+    return false;
+  }
+
+  // for the `armPointLike` call, we need to iterate over all available
+  // processes and select the ones that generate the smallest limits. this is
+  // done using an index-based compile-time recursive call.
+  template <typename generator_t, std::size_t kI0, std::size_t... kIs>
+  void armPointLikeImpl(generator_t& rng, const Particle& particle,
+                        Selection& selection,
+                        std::index_sequence<kI0, kIs...>) const {
+    // only arm the process if it is not masked
+    if (not m_mask[kI0]) {
+      auto [x0Limit, l0Limit] =
+          std::get<kI0>(m_processes).generatePathLimits(rng, particle);
+      if (x0Limit < selection.x0Limit) {
+        selection.x0Limit = x0Limit;
+        selection.x0Process = kI0;
+      }
+      if (l0Limit < selection.l0Limit) {
+        selection.l0Limit = l0Limit;
+        selection.l0Process = kI0;
+      }
```

Comment:
Just checking, this means we select the process that will come *next* in x0/l0 and discard any process that has a longer random limit (for now), right?

Reply:
Yes. If the process gets executed and the particle survives, new limits are drawn to determine the next process.